### ISSUE TITLE
Node.js vs C on x86 vs ARM performance tests

### ISSUE THREAD

**Issue by shodan8192** [nodejs/help#1365]
* **Node.js Version**: v8.11.3
* **OS**: multiple
* **Test code**: Mandelbrot fractal generation, [JS and C source](https://gist.github.com/shodan8192/e425222dd300ca316d06644092b8057c)

I have done some tests on different platforms, to compare Node.js and C performance and here are results :

| ﻿CPU                     | Arch   | OS                   | fastest gcc flag | Node.js time [ms] | C time [ms] | Node.js to C ratio |
|-------------------------|--------|----------------------|------------------|-------------------|-------------|--------------------|
| AMD E-350               | x86    | Ubuntu 18.04 x86     | -Os              | 2771.60           | 2877.39     | 1.04               |
| AMD E-350               | x86-64 | Ubuntu 16.04 x64     | -O3              | 2620.93           | 2134.18     | 0.81               |
| Intel Xeon X5460        | x86-64 | Ubuntu 18.04 x64     | -O1              | 1388.72           | 1352.26     | 0.97               |
| Intel Core i7-7700      | x86-64 | Ubuntu 16.04 x64     | -O1              | 801.44            | 783.99      | 0.98               |
| Broadcom BCM2835@1GHz   | ARMv6  | Raspbian stretch 9.4 | -O3              | 18167.38          | 8378.08     | 0.46               |
| Broadcom BCM2837@900MHz | ARMv7  | Raspbian stretch 9.4 | -O2              | 8132.55           | 3493.45     | 0.43               |
| SAMSUNG Exynos7580      | ARMv8  | Android 8.1.0        | -O3              | 4873.04           | 2258.28     | 0.46               |

On x86 / x86-64 the Node.js performance is excellent, but on both Raspberries, it's only about half of C speed, so my question is : what is cause of this ? ARM architecture or particular implementation of it ?

I have no other ARM boards to do more tests, but maybe some of You could do it on different platforms ?
 
I am interested of the results on powerfull ARMs like Cortex-A57,A72 or A73, which have Out-Of-Order Execution and Branch Prediction, to see if they can compete with x86 when it comes to Node.JS performance.

**Comment by vdeturckheim** [comment]
IMHO, the performance difference is more likelly to be a topic for V8 than for Node.js given what I saw in the tests. Have you tried with V8 only?

**Comment by shodan8192** [comment]
No I haven't tried on pure V8, but I think the results will be similar. You're right, it's rather a matter of v8 performance, but I'm user of Node.js and testing on it is convenient. I don't treat this tests result as issue, but more as curiosity, and I want to find out where the differences come from.

BTW maybe some of RPi 3B users run 64bit OS on it (eg Arch Linux or OpenSUSE), and can check if ARMv8 architecture gives better numbers in JS vs C tests.

**Comment by shodan8192** [comment]
I have opened issue on V8 tracker, for anyone who want to look at it here is link : [7928](https://bugs.chromium.org/p/v8/issues/detail?id=7928).

In short : mandelbrot test turned out be the only, on which it's easy to see how microarchitecture affects performance. On advanced cores like Exynos-M2 or Cortex-A72 Node.js / C ratio is similar like on fast x64, in opposite to ARM11 or Cortex-A53 where is about half of it.

Additionally, in mandelbrot test on all ARMs there is performance boost (7-25%) when using --turbo-instruction-scheduling option, but in other tests I've conducted, the profit is minor - details in link above.

**Comment by gireeshpunathil** [comment]
Very interesting topic as well as the observations. One general observation is that it is short running test (less than a second) so the code does NOT get (JIT) re-compiled. of course, not significant w.r.t your stated observations.

One of the pragmatic approach is to do `--print-code` in the 2 type of systems and see if the sequence differs. If no, your inference becomes the conclusion, or else there could be other considerations that changes the performance. Does this sound sensible to you?

**Comment by shodan8192** [comment]
Test is short running (at least on fast x64) but it's get recompiled by JIT I think :

`[marking 0x10d19df825b1 <JSFunction mandelbrot (sfi = 0x3f2639079039)> for optimized recompilation, reason: hot and stable, ICs with typeinfo: 12/12 (100%), generic ICs: 0/12 (0%)]`

`[compiling method 0x10d19df825b1 <JSFunction mandelbrot (sfi = 0x3f2639079039)> using TurboFan OSR]`

`[optimizing 0x10d19df825b1 <JSFunction mandelbrot (sfi = 0x3f2639079039)> - took 0.963, 0.635, 0.006 ms]`

`[compiling method 0x10d19df825b1 <JSFunction mandelbrot (sfi = 0x3f2639079039)> using TurboFan]`

`[optimizing 0x10d19df825b1 <JSFunction mandelbrot (sfi = 0x3f2639079039)> - took 0.106, 0.459, 0.009 ms]`

on ARMs it looks the same.

<br />

> Is the code generated by JIT the same for eg Cortex-A53 and Cortex-A72 ?

 it was one of mine questions that I asked on V8 tracker, so I was thinking about it. I'm gonna check it myself by method mentioned by You, but I have to get again big ARM device for comparison, and I will share the results.

**Comment by gireeshpunathil** [comment]
thanks for the information. I didn't bother to check the JIT verbose option as I wasn't expecting it to recompile within few millis. Great!

So as the compilation and recompilation profile is same in ARM and x64, the possibilities are narrowing down (IMO) to:

 - generated code being different
 - generated code is same, but calling conventions are not native to the architecture
 - everything is same, but the micro-architectural difference being critical

**Comment by shodan8192** [comment]
I have compared Mandelbrot test [JIT code output](https://gist.github.com/shodan8192/5638cf7f4291dd7aa5329289ed7566c4) for:
- Cortex-A53 (Node.js / C ratio : 0.57)
- Exynos-M2 (Node.js / C ratio : 1.01)

**both codes**, except of course addresses and offsets **are identical**, so it proves my inference about microarchitecture.

I also suppose that JIT output should be the same for x64 pair :
- Intel Atom 330 (Node.js / C ratio : 0.69)
- Intel Core i7-7700 (Node.js / C ratio : 0.98)

but I haven't check that yet.

**Comment by shodan8192** [comment]
My JIT output comparison was about "big" vs "LITTLE" ARMv8-A, so calling convention (although not most efficient for ARM64) does not matter here - for both cores was the same. That comparison shows how given ARM microarch affects JITed code vs compiled C performance.

Now, let's back to x64 vs ARM64. Here I agree that non-native calling convention in JIT generated code for ARM64 have to be taken into account as disadvantage. Should that be considered as issue in V8 JIT compiler ?

**Comment by gireeshpunathil** [comment]
> Should that be considered as issue in V8 JIT compiler ?

Yes. While one can argue that the cost for true porting the `codegen` native to each architecture can outweigh the performance gain, that is a one time cost while the perf gain is pervasive, across workloads. 

caveat: the magnitude of gain is something which I don't know but is going to be a function of density of computations (more register usage as opposed to memory) and frequency of subroutine invocations (better leverage of SP, LR R0 - R12).

**Comment by shodan8192** [comment]
I like your explanations. In my spare time, I will try to estimate or measure how much overhead is introduced by passing parameters via stack instead of registers on the Mandelbrot code.

**Comment by gireeshpunathil** [comment]
@shodan8192 - while we head towards closing this issue, I would like to invite you to contribute to Node.js project. Depending on your taste and skills there are variety of areas you can participate (https://github.com/nodejs/benchmarking looks to be a natural fit) but there could be others. Just a suggestion!

**Comment by shodan8192** [comment]
It's very kind of You, thanks !

It was not so long ago, when I became user of Node.js and many things are new for me like that what happens "under the hood" for example. My skill, that I like most is to dig into problem, not having much particular knowledge, but discovering it on the way. If in future I will have something valuable for Node.js I will share it for sure.

I wish You all the best !

**Comment by shodan8192** [comment]
> I will try to estimate or measure how much overhead is introduced by passing parameters via stack instead of registers on the Mandelbrot code.

Quick and dirty test: I modified Mandelbrot C program adding 8 dummy parameters, and decreasing MAX_ITERATIONS to 10.

I compared two cases :
- x and y passed in registers :
```c
int mandelbrot(double x, double y, ... , double g, double h)
```

- x and y passed on stack :
```c
int mandelbrot(double a, double b, ... ,  double x, double y)
```
which adds 2 instructions to function body
```asm
ldr d6, [sp]
ldr d5, [sp, 8]
```
Also 8 instructions setting dummy parameters were added to loop in main function in both cases, so they not impact the result.

Code get **6% slower** but ... it is when function call rate is about **7 millions calls per second** and total execution time is less than **1ms**.

It may be that it's really not worth to perfect tune calling convention on each architecture, having one universal, that will fit (with minor changes) to all of them and make it easier portable.
